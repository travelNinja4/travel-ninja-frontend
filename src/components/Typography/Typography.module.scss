/**
------------------------------------------------------
styles specific to the Typography component
------------------------------------------------------
**/

.typography {
  margin: 0;
  padding: 0;
  font-family: var(--font-inter);
  color: #000000;

  // Alignment
  &.align-left {
    text-align: left;
  }
  &.align-center {
    text-align: center;
  }
  &.align-right {
    text-align: right;
  }
  &.align-right {
    text-align: justify;
  }
  &.align-right {
    text-align: start;
  }
  &.align-right {
    text-align: end;
  }
  &.align-right {
    text-align: match-parent;
  }

  // Text Transform
  &.transform-uppercase {
    text-transform: uppercase;
  }
  &.transform-lowercase {
    text-transform: lowercase;
  }
  &.transform-capitalize {
    text-transform: capitalize;
  }

  // Truncation
  &.ellipsis {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &.noWrap {
    white-space: nowrap;
  }

  // Line Clamping
  @for $i from 1 through 5 {
    &.lineClamp-#{$i} {
      display: -webkit-box;
      -webkit-line-clamp: #{$i};
      -webkit-box-orient: vertical;
      overflow: hidden;
    }
  }
}
