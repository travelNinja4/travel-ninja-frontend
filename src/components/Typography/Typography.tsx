/**
 * custom typography component that standardizes text styles across the application.
 *
 * @example
 * ```tsx
 * import Typography from '@src/components/Typography'
 *
 * export default function Typography() {
 *   return <Typography label="Hello" />;
 * }
 * ```
 */

import React, { JSX, ReactNode, ElementType } from 'react';
import clsx from 'clsx';
import styles from './Typography.module.scss';

/**
 * Define the props available for the Typography component.
 */
export interface TypographyProps {
  tag: Tags;
  as?: ElementType;
  children?: ReactNode;
  className?: string;
  align?: TextAlign;
  transform?: TextTransform;
  truncation?: Truncation;
}

type TextTransform = 'uppercase' | 'lowercase' | 'capitalize' | 'none';

type Truncation = 'ellipsis' | 'noWrap' | `lineClamp-${number}`;

type TextAlign = 'left' | 'right' | 'center' | 'justify' | 'start' | 'end' | 'match-parent';

type Tags = 'h1' | 'h2' | 'h3' | 'h4' | 'p' | 'h6' | 'span';

const TagsMapping: Record<Tags, keyof JSX.IntrinsicElements> = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h6: 'h6',
  p: 'p',
  span: 'span',
};

export default function Typography({
  tag,
  as,
  children,
  className,
  align = 'left',
  transform = 'none',
  truncation,
}: TypographyProps) {
  const Component = as || TagsMapping[tag] || 'p';
  return (
    <Component
      data-testid="TypographyTest"
      className={clsx(
        styles.typography,
        styles[tag],
        styles[`align-${align}`],
        transform !== 'none' && styles[`transform-${transform}`],
        truncation && styles[truncation],
        className,
      )}
    >
      {children}
    </Component>
  );
}
