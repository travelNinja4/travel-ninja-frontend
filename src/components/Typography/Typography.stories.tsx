import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import Typography from './Typography';

const meta = {
  title: 'components/Typography',
  component: Typography,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    tag: {
      control: 'select',
      options: ['h1', 'h2', 'h3', 'h4', 'h6', 'p', 'span'],
      description: 'The semantic HTML tag for the typography element',
    },
    as: {
      control: 'select',
      options: ['p', 'span', 'div', 'label', 'strong', 'em'],
      description: 'Overrides the rendered HTML element while keeping the styling of `tag`',
    },
    children: {
      control: 'text',
      description: 'The text content or elements inside Typography',
    },
    align: {
      control: 'select',
      options: ['left', 'center', 'right', 'justify', 'start', 'end', 'match-parent'],
      description: 'Text alignment',
    },
    transform: {
      control: 'select',
      options: ['none', 'uppercase', 'lowercase', 'capitalize'],
      description: 'Text transformation style',
    },
    truncation: {
      control: 'select',
      options: [
        'ellipsis',
        'noWrap',
        'lineClamp-1',
        'lineClamp-2',
        'lineClamp-3',
        'lineClamp-4',
        'lineClamp-5',
      ],
      description: 'Truncates or clamps text display',
    },
    className: {
      control: 'text',
      description: 'Custom class name for additional styling',
    },
  },
  args: {
    tag: 'p',
    children: 'This is a sample typography text.',
    align: 'left',
    transform: 'none',
    truncation: undefined,
  },
} satisfies Meta<typeof Typography>;

export default meta;
type Story = StoryObj<typeof meta>;

// Default story
export const Default: Story = {};

// Show heading variants
export const Headings: Story = {
  render: (args) => (
    <>
      <Typography {...args} tag="h1">
        Heading 1
      </Typography>
      <Typography {...args} tag="h2">
        Heading 2
      </Typography>
      <Typography {...args} tag="h3">
        Heading 3
      </Typography>
      <Typography {...args} tag="h4">
        Heading 4
      </Typography>
      <Typography {...args} tag="h6">
        Heading 6
      </Typography>
    </>
  ),
};

// Show alignments
export const Alignments: Story = {
  render: (args) => (
    <>
      <Typography {...args} align="left">
        Left aligned text
      </Typography>
      <Typography {...args} align="center">
        Center aligned text
      </Typography>
      <Typography {...args} align="right">
        Right aligned text
      </Typography>
      <Typography {...args} align="justify">
        Justify aligned text
      </Typography>
      <Typography {...args} align="start">
        Start aligned text
      </Typography>
      <Typography {...args} align="end">
        End aligned text
      </Typography>
      <Typography {...args} align="match-parent">
        Match-Parent aligned text
      </Typography>
    </>
  ),
};

// Show text transforms
export const Transforms: Story = {
  render: (args) => (
    <>
      <Typography {...args} transform="uppercase">
        Uppercase text
      </Typography>
      <Typography {...args} transform="lowercase">
        Lowercase Text
      </Typography>
      <Typography {...args} transform="capitalize">
        capitalize text
      </Typography>
    </>
  ),
};

// Show truncation styles
export const Truncations: Story = {
  render: (args) => (
    <div style={{ width: '250px' }}>
      <Typography {...args} truncation="ellipsis">
        This is a very long sentence that will be cut off with an ellipsis.
      </Typography>
      <Typography {...args} truncation="noWrap">
        This is a long text that will not wrap at all.
      </Typography>
      <Typography {...args} truncation="lineClamp-2">
        This is a long paragraph that will be clamped to two lines in height, and anything beyond
        that will be hidden from view.
      </Typography>
    </div>
  ),
};

// Playground with full controls
export const Playground: Story = {
  args: {
    tag: 'p',
    as: 'p',
    children: 'Edit me in Storybook controls!',
    align: 'left',
    transform: 'none',
    truncation: undefined,
  },
};
