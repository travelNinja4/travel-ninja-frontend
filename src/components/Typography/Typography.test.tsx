import '@testing-library/jest-dom';
import { ElementType } from 'react';
import { render, screen } from '@testing-library/react';
import Typography, { TypographyProps } from './Typography';

describe('Home', () => {
  /** Base props for Typography **/
  const baseProps: TypographyProps = {
    tag: 'p',
    children: 'Typography Test',
    align: 'left',
  };

  /**
   * @test
   * Renders the Typography component without throwing an error
   * and checks if the TypographyTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<Typography {...baseProps} />);
    expect(screen.getByTestId('TypographyTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the Typography component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<Typography {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the Typography component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<Typography {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders using the "as" prop instead of tag', () => {
    render(
      <Typography {...baseProps} as="strong">
        Strong text
      </Typography>,
    );
    const el = screen.getByText('Strong text');
    expect(el.tagName.toLowerCase()).toBe('strong');
  });

  it.each(['uppercase', 'lowercase', 'capitalize'] as const)(
    'applies transform class for %s',
    (transform) => {
      render(<Typography {...baseProps} transform={transform} />);
      expect(screen.getByTestId('TypographyTest')).toHaveClass(`transform-${transform}`);
    },
  );

  it.each(['ellipsis', 'noWrap', 'lineClamp-3'] as const)(
    'applies truncation class for %s',
    (truncation) => {
      render(<Typography {...baseProps} truncation={truncation} />);
      expect(screen.getByTestId('TypographyTest')).toHaveClass(truncation);
    },
  );

  it('falls back to <p> tag if tag is missing', () => {
    render(
      <Typography
        as={undefined as ElementType | undefined}
        tag={undefined as unknown as TypographyProps['tag']}
      >
        Fallback text
      </Typography>,
    );
    const el = screen.getByText('Fallback text');
    expect(el.tagName.toLowerCase()).toBe('p');
  });
});
