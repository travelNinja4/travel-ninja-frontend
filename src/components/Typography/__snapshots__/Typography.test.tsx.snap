// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="typography p align-left"
    data-testid="TypographyTest"
  >
    Typography Test
  </p>
</DocumentFragment>
`;
